src/realm/object-store/sectioned_results.cpp

```diff
@@ -0,0 +1,385 @@
+            for (auto& [hash, index] : b) {
+                if (a.find(hash) == a.end()) {
+                    out.add(b.at(hash));
+                }
+            }
+        };
+
+        diff(old_sections, new_sections, sections_to_insert);
+        diff(new_sections, old_sections, sections_to_remove);
+
+        return {sections_to_insert, sections_to_remove};
+    }
+
+private:
+    SectionedResultsNotificatonCallback m_cb;
+    SectionedResults& m_sectioned_results;
+    std::vector<Section> m_prev_sections;
+    std::vector<std::pair<size_t, size_t>> m_prev_row_to_index_path;
+    util::Optional<size_t> m_section_filter;
+};
+
+Mixed ResultsSection::operator[](size_t idx) const
+{
+    m_parent->calculate_sections_if_required();
+    return m_parent->m_results.get_any(m_parent->m_sections[m_index].indices[idx]);
+}
+
+Mixed ResultsSection::key()
+{
+    return m_parent->m_sections[m_index].key;
+}
+
+size_t ResultsSection::size()
+{
+    m_parent->calculate_sections_if_required();
+    REALM_ASSERT(m_parent->m_sections.size() > m_index);
+    auto range = m_parent->m_sections[m_index];
+    return range.indices.size();
+}
+
+NotificationToken ResultsSection::add_notification_callback(SectionedResultsNotificatonCallback callback,
+                                                            KeyPathArray key_path_array) &
+{
+    return m_parent->add_notification_callback_for_section(m_index, std::move(callback), key_path_array);
+}
+
+realm::ThreadSafeReference ResultsSection::thread_safe_reference()
+{
+    return m_parent->thread_safe_reference();
+}
+
+bool ResultsSection::is_valid() const
+{
+    return m_parent->is_valid();
+}
+
+size_t ResultsSection::hash() const
+{
+    return m_parent->m_sections[m_index].hash;
+}
+
+SectionedResults::SectionedResults(Results results, SectionKeyFunc section_key_func)
+    : m_results(results)
+    , m_callback(std::move(section_key_func))
+{
+    calculate_sections_if_required(true);
+}
+
+SectionedResults::SectionedResults(Results results, Results::SectionedResultsOperator op,
+                                   util::Optional<StringData> prop_name)
+    : m_results(results)
+    , m_callback(builtin_comparison(results, op, prop_name))
+{
+    calculate_sections_if_required(true);
+}
+
+void SectionedResults::calculate_sections_if_required(bool force_update)
+{
+    if (m_results.m_update_policy == Results::UpdatePolicy::Never)
+        return;
+    else if (!m_results.has_changed() && !force_update)
+        return;
+
+    {
+        util::CheckedUniqueLock lock(m_results.m_mutex);
+        m_results.ensure_up_to_date();
+    }
+
+    calculate_sections();
+}
+
+// This method will run in the following scenarios:
+// - SectionedResults is freshly created from Results
+// - The underlying Table in the Results collection has changed
+void SectionedResults::calculate_sections()
+{
+    auto r = m_results.snapshot();
```

Comment:
I don't see any obvious reason this needs a snapshot?

Reply:
I was unsure if this was actually needed, my thought was that we need a stable snapshot of the results incase they change while the loop is running.

Reply:
I have removed the need for the snapshot